### PR TITLE
Add `YEARWEEK` Function To OpenSearch SQL

### PR DESCRIPTION
**Pull request by GabeFernandez310**
### Description
The `yearweek(date [, mode])` function accepts a first argument representing a date, and an optional second argument containing a mode. It extract the **year** and **week of the year** from the input argument, and formats it as an integer which is returned. If an argument of type `TIME` is provided, the current date is used. The mode argument defines a set of rules to decide how to parse the **year** and **week of the year** and are based on [MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_yearweek).

Example:

```
opensearchsql> SELECT YEARWEEK('2020-08-26');
fetched rows / total rows = 1/1
+--------------------------+
| YEARWEEK('2020-08-26')   |
|--------------------------|
| 202034                   |
+--------------------------+
```

```
opensearchsql> SELECT YEARWEEK('2020-08-26', 0);
fetched rows / total rows = 1/1
+-----------------------------+
| YEARWEEK('2020-08-26', 0)   |
|-----------------------------|
| 202034                      |
+-----------------------------+
```
 
### Issues Resolved
[722](https://github.com/opensearch-project/sql/issues/722)
 
### Check List
- [X] New functionality includes testing.
  - [X] All tests pass, including unit test, integration test and doctest
- [X] New functionality has been documented.
  - [X] New functionality has javadoc added
  - [X] New functionality has user manual doc added
- [X] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 1]
```diff
@@ -900,6 +901,30 @@ private DefaultFunctionResolver year() {
     );
   }
 
+  /**
+   * YEARWEEK(DATE[,mode]). return the week number for date.
```
Comment (matthewryanwells) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:905]:
```suggestion
   * YEARWEEK(DATE,[mode]). return the week number for date.
```

Comment (acarbonetto) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:905]: The syntax is correct as is.  If the mode isn't included, the comma should not be included. This suggestion will say that this is correct syntax:
```
YEARWEEK('12-12-12', )
```

**Review by acarbonetto — APPROVED** [review]
LGTM

**Review by matthewryanwells — APPROVED** [review]

**Review by matthewryanwells — APPROVED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java [hunk 2]
```diff
@@ -1731,6 +1756,52 @@ private ExprValue exprYear(ExprValue date) {
...
+
+  /**
+   * Yearweek for date implementation for ExprValue.
+   * When mode is not specified default value mode 0 is used.
+   *
+   * @param date ExprValue of Date/Datetime/Time/Timestamp/String type.
+   * @return ExprValue.
+   */
+  private ExprValue exprYearweekWithoutMode(ExprValue date) {
+    return exprYearweek(date, new ExprIntegerValue(0));
```
Comment (Yury-Fridlyand) [inline core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1798]:
```suggestion
    return exprYearweek(date, DEFAULT_WEEK_OF_YEAR_MODE);
```

**Review by Yury-Fridlyand — DISMISSED** [review]

File: core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java (hunk 2 above)
Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1798]: This was done on purpose to try an match [MySQL](https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html#function_yearweek:~:text=Unlike%20WEEK()%2C%20the%20value%20of%20default_week_format%20does%20not%20influence%20YEARWEEK().). I can use this suggestion though if that is what is preferred.

Comment (Yury-Fridlyand) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1798]: It is already used in this function, I just propose to use this constant everywhere.

Comment (GabeFernandez310) [reply core/src/main/java/org/opensearch/sql/expression/datetime/DateTimeFunction.java:1798]: Fixed in https://github.com/opensearch-project/sql/pull/1417/commits/d4fba473572256cded247cdebb91777392b7ed02

**Review by Yury-Fridlyand — APPROVED** [review]

**Review by dai-chen — APPROVED** [review]

**Comment by Yury-Fridlyand** [comment]
@GabeFernandez310 please, backport manually

**Pull request merged**